### PR TITLE
fix(sort): Make sort consistent for indexed and without indexed predicates

### PR DESCRIPTION
**Pull request by ahsanbarkati**
This PR makes the result of sort consistent for predicate with and without index. There was an issue that for predicate with index, we used to drop the nodes which doesn't contain the predicate used for sorting, while we retained them for that of without index. 

The nulls were dropped in thee case of `sortWithIndex` because in this case to calculate the result, we used to take intersection of the given list for sorting with the index of the predicate used for sorting, hence other predicates were dropped. In this change, we keep track of nodes which have null sort predicates and append them to the result as required.

Some benchmarks:
The dataset contain 1million UIDs, with half of them containing the predicate used for sorting. The time taken are in nano seconds.
```
No Index:
Query                                   Time
orderasc, first:100 ->                  7092526995
orderasc, offset:100000, first:100 ->   6809554104
orderdesc, first:100 ->                 6981851622
orderdesc, offset:100000, first:100 ->  6862242806

Index:
orderasc, first:100 ->                  1421648793
orderasc, offset:100000, first:100 ->   2044206032
orderdesc, first:100 ->                 1577750157
orderdesc, offset:100000, first:100 ->  1956177772
```

The results of the same queries on **Master**:
```
No Index:
orderasc, first:100 ->                  6881212989
orderasc, offset:100000, first:100 ->   7543954929
orderdesc, first:100 ->                 7327315514
orderdesc, offset:100000, first:100 ->  7262242806

Index:
orderasc, first:100 ->                  1410905825
orderasc, offset:100000, first:100 ->   1890659562
orderdesc, first:100 ->                 1404256406
orderdesc, offset:100000, first:100 ->  2106243131
```

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7241)

### CODE REVIEW

**Comment by vvbalaji-dgraph** [comment]
@danielmai : this change is expected to produce consistent results while retaining performance of using sorted index.

File: worker/sort.go [hunk 1]
```diff
@@ -302,6 +305,44 @@ BUCKETS:
...
+			if _, ok := present[uid]; !ok {
+				toAppend = append(toAppend, uid)
+			}
+		}
+		reqCount := int(ts.Count) - len(r.UidMatrix[i].Uids)
+		fmt.Println("result", r.UidMatrix[i].Uids)
+		fmt.Println("skipped", out[i].skippedList.Uids)
+		fmt.Println("toappend", toAppend)
+		if order.Desc {
+			r.UidMatrix[i].Uids = append(out[i].skippedList.Uids, r.UidMatrix[i].Uids...)
```
Comment (ahsanbarkati) [inline worker/sort.go:328]: Do we need to append skippedList for descending case?

**Review by vmrajas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7241)* status: 0 of 5 files reviewed, 3 unresolved discussions (waiting on @ahsanbarkati, @manishrjain, @pawanrawal, @vmrajas, and @vvbalaji-dgraph)

___
*[worker/sort.go, line 185 at r5](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQk3idp0pAInD6TsHaG:-MQk3idp0pAInD6TsHaH:b-rhlnwl) ([raw file](https://github.com/dgraph-io/dgraph/blob/74e162af07820b273cce8d9ddb8396cd6bf6f80a/worker/sort.go#L185)):*
> ```Go
> 	span.Annotate(nil, "sortWithIndex")
> 
> 	maxCount := 0
> ```

Can you add a comment on what maxCount is going to store.
___
*[worker/sort.go, line 287 at r5](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQk3r7tD3Ol34fkydnt:-MQk3r7tD3Ol34fkydnu:b19etit) ([raw file](https://github.com/dgraph-io/dgraph/blob/74e162af07820b273cce8d9ddb8396cd6bf6f80a/worker/sort.go#L287)):*
> ```Go
> 			token := k.Term
> 			if !order.Desc {
> 				maxCount = int(ts.Count)
> ```

Can you add a comment on why this is done. I know that this would become a rather large comment. But, it would help in understanding the code later.
___
*[worker/sort.go, line 606 at r5](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQk6OVL8BvhIvqxVQIv:-MQk6OVL8BvhIvqxVQIw:b-8x6q7k) ([raw file](https://github.com/dgraph-io/dgraph/blob/74e162af07820b273cce8d9ddb8396cd6bf6f80a/worker/sort.go#L606)):*
> ```Go
> 
> // intersectBucket intersects every UID list in the UID matrix with the
> // indexed bucket.
> ```

Can you also add a comment on the significance of count parameter.

**Review by pawanrawal — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7241)* status: 0 of 5 files reviewed, 10 unresolved discussions (waiting on @ahsanbarkati, @manishrjain, and @vvbalaji-dgraph)

___
*[query/query1_test.go, line 1963 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQuJiP7AIqntk19X-He:-MQuJiP8AqHcqAI8oUp_:b-cfkn7j) ([raw file](https://github.com/dgraph-io/dgraph/blob/deb4b7966f4baedd457cfcd14d17efeb6b5f895f/query/query1_test.go#L1963)):*
> ```Go
> }
> 
> func TestSortNull2(t *testing.T) {
> ```

All these could just have been one test with Table driven tests because you are testing the same thing that is null behavior with various values of first and offset
___
*[query/query1_test.go, line 2095 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQuJuDSBtyFPtve7uZv:-MQuJuDSBtyFPtve7uZw:bhaqk95) ([raw file](https://github.com/dgraph-io/dgraph/blob/deb4b7966f4baedd457cfcd14d17efeb6b5f895f/query/query1_test.go#L2095)):*
> ```Go
> 
> 	query := `{
> me(func: uid(61, 62, 63, 64, 65, 66, 67, 68, 69, 70), orderdesc: pred, first: 2) {
> ```

How about a test with `offset:5 and first:5` and another one with `offset:9 and first:5`
___
*[query/query2_test.go, line 977 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQuK7Ac2D7NtjgrkdPs:-MQuK7Ac2D7NtjgrkdPt:b-xuqnp5) ([raw file](https://github.com/dgraph-io/dgraph/blob/deb4b7966f4baedd457cfcd14d17efeb6b5f895f/query/query2_test.go#L977)):*
> ```Go
> 			"data": {
> 				"q": [{
> 					"name_lang_index@de": "öffnen",
> ```

Are these the null value ones?
___
*[worker/sort.go, line 195 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQuKLYuE7MH_yhJHsq3:-MQuKLYuE7MH_yhJHsq4:bl0rl78) ([raw file](https://github.com/dgraph-io/dgraph/blob/deb4b7966f4baedd457cfcd14d17efeb6b5f895f/worker/sort.go#L195)):*
> ```Go
> 		var emptySkippedList pb.List
> 		out[i].ulist = &emptyList
> 		out[i].skippedUids = &emptySkippedList
> ```

Just define it inline like
```
out[i].skippedUids = &pb.List{}
```
___
*[worker/sort.go, line 325 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQuKhUP7kxzQRtORFV7:-MQuKhUP7kxzQRtORFV8:bjta4q6) ([raw file](https://github.com/dgraph-io/dgraph/blob/deb4b7966f4baedd457cfcd14d17efeb6b5f895f/worker/sort.go#L325)):*
> ```Go
> 		for _, uid := range ul.Uids {
> 			if _, ok := present[uid]; !ok {
> 				nullPreds = append(nullPreds, uid)
> ```

So this `nullPreds` is common across all the lists? Shouldn't it be reset for each uid list?
___
*[worker/sort.go, line 329 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQuL0LP0kLHrWnVt1SU:-MQuL0LP0kLHrWnVt1SV:b-s22840) ([raw file](https://github.com/dgraph-io/dgraph/blob/deb4b7966f4baedd457cfcd14d17efeb6b5f895f/worker/sort.go#L329)):*
> ```Go
> 		}
> 
> 		requiredCount := int(ts.Count) - len(r.UidMatrix[i].Uids)
> ```

`remainingCount` would be a better name.
___
*[worker/sort.go, line 769 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQuLdZT6aCn2VIDUTIO:-MQuLdZT6aCn2VIDUTIP:b-rmd9fu) ([raw file](https://github.com/dgraph-io/dgraph/blob/deb4b7966f4baedd457cfcd14d17efeb6b5f895f/worker/sort.go#L769)):*
> ```Go
> 				val.Value = nil
> 				nullsList = append(nullsList, uid)
> 				nullVals = append(nullVals, []types.Val{val})
> ```

What does `val` hold? Is it just a null? Would be the `null` be returned? I don't see `null` being returned in the query tests.

**Review by ahsanbarkati — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7241)* status: 0 of 5 files reviewed, 10 unresolved discussions (waiting on @ahsanbarkati, @manishrjain, @pawanrawal, and @vvbalaji-dgraph)

___
*[query/query1_test.go, line 1963 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQuJiP7AIqntk19X-He:-MQunONj3UfCf4hO2jbh:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/deb4b7966f4baedd457cfcd14d17efeb6b5f895f/query/query1_test.go#L1963)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

All these could just have been one test with Table driven tests because you are testing the same thing that is null behavior with various values of first and offset
</blockquote></details>

Done.
___
*[query/query1_test.go, line 2095 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQuJuDSBtyFPtve7uZv:-MQunQeRBYZmNxdZeyYD:bi6isgd) ([raw file](https://github.com/dgraph-io/dgraph/blob/deb4b7966f4baedd457cfcd14d17efeb6b5f895f/query/query1_test.go#L2095)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

How about a test with `offset:5 and first:5` and another one with `offset:9 and first:5`
</blockquote></details>

Added these cases.
___
*[query/query2_test.go, line 977 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQuK7Ac2D7NtjgrkdPs:-MQunSki1hD-QYL9RSDM:bb77ufg) ([raw file](https://github.com/dgraph-io/dgraph/blob/deb4b7966f4baedd457cfcd14d17efeb6b5f895f/query/query2_test.go#L977)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Are these the null value ones?
</blockquote></details>

Yes.
___
*[worker/sort.go, line 195 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQuKLYuE7MH_yhJHsq3:-MQunU-WCUPPmLWfdX4w:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/deb4b7966f4baedd457cfcd14d17efeb6b5f895f/worker/sort.go#L195)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Just define it inline like
```
out[i].skippedUids = &pb.List{}
```
</blockquote></details>

Done.
___
*[worker/sort.go, line 325 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQuKhUP7kxzQRtORFV7:-MQunVQC8rc5npxR1j_u:b9cql6i) ([raw file](https://github.com/dgraph-io/dgraph/blob/deb4b7966f4baedd457cfcd14d17efeb6b5f895f/worker/sort.go#L325)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

So this `nullPreds` is common across all the lists? Shouldn't it be reset for each uid list?
</blockquote></details>

Thanks. Fixed.
___
*[worker/sort.go, line 329 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQuL0LP0kLHrWnVt1SU:-MQunYb27mseCtOxqdQs:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/deb4b7966f4baedd457cfcd14d17efeb6b5f895f/worker/sort.go#L329)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

`remainingCount` would be a better name.
</blockquote></details>

Done.
___
*[worker/sort.go, line 769 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQuLdZT6aCn2VIDUTIO:-MQunZqvB1Sf-FhdHWB7:b-iln05h) ([raw file](https://github.com/dgraph-io/dgraph/blob/deb4b7966f4baedd457cfcd14d17efeb6b5f895f/worker/sort.go#L769)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

What does `val` hold? Is it just a null? Would be the `null` be returned? I don't see `null` being returned in the query tests.
</blockquote></details>

Yes. It is just the null values. It is being used for multisort, I wanted to keep the behavior of multisort unchanged, so appended the null values corresponding to the uid list, for which the predicates contains null or unsortable values.

**Review by pawanrawal — CHANGES_REQUESTED** [review]
Do a bit more testing around different offsets and first combinations for different kind of datasets as discussed.

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-:-MQv3I2g5zaGEP3wn1PH:btkn439)* status: 0 of 5 files reviewed, 10 unresolved discussions (waiting on @ahsanbarkati, @manishrjain, @pawanrawal, and @vvbalaji-dgraph)

___
*[query/query1_test.go, line 2008 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQv09YV1nJSZHRFsRZv:-MQv09YV1nJSZHRFsRZw:bjp2aoh) ([raw file](https://github.com/dgraph-io/dgraph/blob/d02405ae6feff04c5493b0d791fba7a9108de705/query/query1_test.go#L2008)):*
> ```Go
> 			{"predname":"nameJ"}]}}`,
> 		},
> 		{11, 9, 5, false, `{"data": {"me":[
> ```

How about 12 and 5, does that return an empty result?
___
*[worker/sort.go, line 307 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQv1Kb0Ds6DjiWcS_zR:-MQv1Kb0Ds6DjiWcS_zS:b-bbce6y) ([raw file](https://github.com/dgraph-io/dgraph/blob/d02405ae6feff04c5493b0d791fba7a9108de705/worker/sort.go#L307)):*
> ```Go
> 
> 	for i, ul := range ts.UidMatrix {
> 		var nullPreds []uint64
> ```

Add a comment
These are uids for which there is value for the predicate.
___
*[worker/sort.go, line 328 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQv-a4I16ylxGfs48MM:-MQv-a4I16ylxGfs48MN:b-4e4bz) ([raw file](https://github.com/dgraph-io/dgraph/blob/d02405ae6feff04c5493b0d791fba7a9108de705/worker/sort.go#L328)):*
> ```Go
> 
> 		// We didn't get anything in the intersected result, it is possible that complete offset
> 		// is yet not applied. We need to apply the remainng offset on the null values.
> ```

remaining
___
*[worker/sort.go, line 331 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQv28f8EYM5IiHjp97R:-MQv28f8EYM5IiHjp97S:brftz37) ([raw file](https://github.com/dgraph-io/dgraph/blob/d02405ae6feff04c5493b0d791fba7a9108de705/worker/sort.go#L331)):*
> ```Go
> 		if len(r.UidMatrix[i].Uids) == 0 {
> 			start := int(ts.Offset) - len(out[i].skippedUids.Uids)
> 			x.AssertTrue(start >= 0)
> ```

return an error here instead of crashing the server and think of a case where this doesn't hold
___
*[worker/sort.go, line 335 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQv3BTdCxPKqE2AnQet:-MQv3BTdCxPKqE2AnQeu:ba2cx9b) ([raw file](https://github.com/dgraph-io/dgraph/blob/d02405ae6feff04c5493b0d791fba7a9108de705/worker/sort.go#L335)):*
> ```Go
> 				nullPreds = nullPreds[start:]
> 			} else {
> 				nullPreds = []uint64{}
> ```

nullsPreds = nullPreds[:0]
___
*[worker/sort.go, line 338 at r7](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQv0JK8Em0EV2ZqpdXL:-MQv0JK8Em0EV2ZqpdXM:b-5yuf93) ([raw file](https://github.com/dgraph-io/dgraph/blob/d02405ae6feff04c5493b0d791fba7a9108de705/worker/sort.go#L338)):*
> ```Go
> 			}
> 		}
> 		remainingCount := int(ts.Count) - len(r.UidMatrix[i].Uids)
> ```

Try removing the index on `pred` and your tests should still pass.

**Review by pawanrawal — APPROVED** [review]
> Reviewed 1 of 5 files at r3, 1 of 3 files at r6, 3 of 3 files at r8.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7241)* status: all files reviewed, 11 unresolved discussions (waiting on @ahsanbarkati, @manishrjain, @pawanrawal, and @vvbalaji-dgraph)

___
*[query/query1_test.go, line 2020 at r8](https://reviewable.io/reviews/dgraph-io/dgraph/7241#-MQviu_HEIahL1XUtbnZ:-MQviu_HEIahL1XUtbn_:b-j2w6nl) ([raw file](https://github.com/dgraph-io/dgraph/blob/40aeb532d1c38807188365588edbd258e2bfaf14/query/query1_test.go#L2020)):*
> ```Go
> 
> 	makeQuery := func(offset, first int32, desc, index bool) string {
> 		pred := "pred"
> ```

You can have two queries with the same alias

**Pull request merged**